src/librustc/middle/ty/util.rs

```diff
@@ -338,8 +299,7 @@ impl<'tcx> ty::ctxt<'tcx> {
         match const_eval::eval_const_expr_partial(self, count_expr, hint, None) {
             Ok(val) => {
                 let found = match val {
-                    ConstVal::Uint(count) => return count as usize,
-                    ConstVal::Int(count) if count >= 0 => return count as usize,
+                    ConstVal::Integral(ConstInt::Usize(count)) => return count.as_usize(self),
```

Comment:
I find that too magical. In this case I want to report an error if it's not a usize, the `as_usize` call is because the actual usize might be 32 bit or 64 bit depending on the *target* platform